### ISSUE TITLE
Concurrency flag  performance

### ISSUE THREAD

**Issue by Othmane796** [peak/s5cmd#418]
Hi,

I'm testing the concurrency flag of s5cmd, after a series of test I'm starting to have doubts if the flag is taken into account or not.

I'm copying a file of 20GB from S3 to local folder. The file I have been testing with is this one : 's3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar' (part of the AWS registry of open data).
 
**System detail:**
-Infrastructure : EC2 instance of type c6a.8xlarge (32 CPUs, 64 GiB of RAM and 12500 Megabit of network). More details [here](https://aws.amazon.com/ec2/pricing/on-demand/)
-System: Ubuntu18.04 using the ubuntu AMI (Ubuntu Server 18.04 LTS (HVM), SSD Volume Type)
-s5cmd installed through a packaged release:
 $ curl -OL https://github.com/peak/s5cmd/releases/download/v1.4.0/s5cmd_1.4.0_Linux-64bit.tar.gz
 $ tar -C /s5cmd -xvf s5cmd_1.4.0_Linux-64bit.tar.gz &&\
 $ export PATH="${PATH}:/s5cmd/"

**Commands and their runtime:**

$time aws s3 cp s3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar .

real    **2m32.663s**
user    1m14.195s
sys     0m54.963s

$time s5cmd --log 'debug' cp -c 32 s3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar .
real    **3m26.686s**
user    0m15.998s
sys     5m17.360s

$time s5cmd --log 'debug' cp -c 16 s3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar .
real    **3m29.332s**
user    0m20.657s
sys     2m19.601s

$time s5cmd --log 'debug' cp -c 8 s3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar .

real    **3m32.870s**
user    0m16.246s
sys     0m47.369s

I'm also checking in parallel the stats from "$htop" through the terminal and I see a difference when using "-c 8":

![8-concurent](https://user-images.githubusercontent.com/8608330/161103486-c2856b35-1704-49fd-ada1-06d3984153c5.png)

and when using the "-c 32" flag:

![32-concurent](https://user-images.githubusercontent.com/8608330/161103541-d92d316d-8b2a-49d1-a79f-29222433f710.png)

In short, I'm a bit surprised that 'aws s3 cp' is a bit faster, but mostly by the fact that changing the concurrent flag doesn't impact the execution time of the s5cmd command. Am I doing something wrong?

Thanks

**Comment by Kucukaslan** [comment]
Hi,
I've tried to reproduce this behaviour in current stable version (v2.0.0) but I failed (to reproduce). Could you please verify that this issue still persists? 

I couldn't use [your sample file](s3://aev-autonomous-driving-dataset/camera_lidar-20190401121727_lidar_frontcenter.tar) since it gave "Error: ... Access Denied status code: 403"

`time s5cmd  cp -c 32  s3://SOMEBUCKET/mck/10GB  /tmp/ `
>cp s3://SOMEBUCKET/mck/10GB /tmp/10GB
real	0m15.973s
user	0m5.783s
sys	1m42.939s
><img width="797" alt="Screen Shot 2022-08-08 at 15 44 18" src="https://user-images.githubusercontent.com/32893303/183421117-3d3bc1d4-1815-4d02-ad9a-2dc315b1e438.png">

`time s5cmd  cp -c 8  s3://SOMEBUCKET/mck/10GB  /tmp/`
>cp s3://SOMEBUCKET/mck/10GB /tmp/10GB
real	0m23.559s
user	0m6.086s
sys	0m35.244s
![Screen Shot 2022-08-08 at 15 22 51](https://user-images.githubusercontent.com/32893303/183419463-4da77ac6-7dac-4022-9f69-9d0b1246edbc.png)


ps. I've looked through the changelog but didn't see anything that could've fixed this, maybe sdk upgrade?

**Comment by Kucukaslan** [comment]
Hi,

I've reproduced this behaviour. I'm not exactly sure why it happens and how can it be solved. My observations are

### 1. Concurrency parameters are used.
I enabled the Disk Write (IO) columns in htop. Number of threads that writes to disk depend on concurrency count. There are much more threads doing Disk writes when concurrency is set to 32 than when it is set to 5.

* Concurrency is set to 32
 <img width="1379" alt="Concurrency set to 5" src="https://user-images.githubusercontent.com/32893303/183434335-2debacb0-6abc-4e5a-bcce-5e1d509fcd01.png">
* Concurrency is set to 5
 <img width="1379" alt="Concurrency set to 5" src="https://user-images.githubusercontent.com/32893303/183434030-aafd18c2-9881-4375-9790-112e897fd494.png">


### 2. Unexpected write stall before completion
Just before completion of downloads the disk writes stops for a while (~20 seconds) then it continues and completes within a few seconds. I don't know why. 


### 3. too much concurrency may actualy be causing low disk utilization
Random I/O is slower than the sequential I/O. So too much concurrency might be causing too much page faults thus performs worse than single thread operation. I vaguely remember some of the concepts exploring/explaining this e.g. [trashing](https://www.os-book.com/OS9/slide-dir/PPT-dir/ch9.ppt).

### 4. 
when I tested with s5cmd setting concurrency to 1 was slower than 2 which is in turn slower than 5. But the difference between 5 and 32 was not significant.

### (inconclusive) Conclusion
In my previous answer I used a file of size 10GB.  It was downloaded in ~20 seconds. But when I tried to download the tar file you mentioned (~24 GB) it took more than 3 minutes (+180 seconds). I used a device with 32 GB RAM.  The 10GB file can easily fit into RAM, while the 24GB file cannot. So I guess it is related to increasing number of [page faults](https://en.wikipedia.org/wiki/Page_fault) and so the lower throughput of Random I/O over sequential I/O.
I do not, yet, know how can we optimize it for best performance/throughput (except manually testing to find optimum value of).

**Comment by Kucukaslan** [comment]
For the record, at some time in last month we made benchmark tests using [storage optimized i3 instances](https://aws.amazon.com/ec2/instance-types/i3/) where s5cmd achieved much higher throughput. So using those instances might help to increase download speeds.

Also I made a few attempts to optimize write requests to achieve increase throughput without using the storage optimized instances. But I couldn't find a viable solution. 
Rest of the comment is about my attempts.

I've written a simple go script to test whether the 'buffer/block size' affects the write throughput [here](https://gist.github.com/Kucukaslan/dfdd507f6fccb0abff0647f8e4d6732c#file-diskio-go). In my computer(s) setting `blockSize` to 32 MB vs 32 KB vs 1 KB made significant difference in throughput (32 MB was much faster than others). However when I tested it on EC2 or Google Cloud compute I couldn't observe a stable difference.

Nevertheless I've also written another (standalone) go script which downloads an object from AWS S3. It is available in [this repository](https://github.com/Kucukaslan/PoCSequentiality). Unfortunately my network connection is not fast enough to test download speed accurately. So I tested downloading same file in an GC E2 instance with 12 vCPU and 32 GB RAM, 100 GB disk. However download speed was 2.30 minute at best but it took over 10 minutes for few times.

So it seems my attempts couldn't produce a viable solution, and, at least for now, changing EC2 instance to a storage optimized instance is the only way to achieve higher throughput.
